README.md
# spring-boot-sse-manager
[![GitHub license](https://img.shields.io/github/license/svg-zentrale/spring-boot-sse-manager)](https://github.com/svg-zentrale/spring-boot-sse-manager/blob/master/LICENSE)
![GitHub Workflow Status](https://img.shields.io/github/workflow/status/svg-zentrale/spring-boot-sse-manager/test)
[![codecov](https://codecov.io/gh/svg-zentrale/spring-boot-sse-manager/branch/master/graph/badge.svg)](https://codecov.io/gh/svg-zentrale/spring-boot-sse-manager)


### About

SSEEmitter wrapped in some goodies for info and keep-alive

Features:
* Heartbeat
* Info / Debug / Error / Done Events

### Setup
 
Releases are published to [bintray jcenter](https://bintray.com/svg-zentrale/maven/spring-boot-sse-manager) and 
[maven central](https://maven-badges.herokuapp.com/maven-central/de.svg/spring-boot-sse-manager) 

![Bintray](https://img.shields.io/bintray/v/svg-zentrale/maven/spring-boot-sse-manager)
![Maven Central](https://img.shields.io/maven-central/v/de.svg/spring-boot-sse-manager)

Maven:

```xml
<dependency>
  <groupId>de.svg</groupId>
  <artifactId>spring-boot-sse-manager</artifactId>
  <version>0.1.14</version>
</dependency>
```

Gradle:

```groovy
implementation 'de.svg:spring-boot-sse-manager:0.1.14'
```

##### Snapshots

You can use snapshot versions through [JitPack](https://jitpack.io):

* Go to [JitPack project page](https://jitpack.io/#marvinosswald/spring-boot-sse-manager)
* Select `Commits` section and click `Get it` on commit you want to use (top one - the most recent)
* Follow displayed instruction: add repository and change dependency (NOTE: due to JitPack convention artifact group will be different)

### Usage

#### Parallel Tasks 

```java
@GetMapping(path = "/stream")
public SseEmitter downloadLink() {
        return new SSEStream((SSEStream stream) -> {
            List<Integer> collect = Stream.of(1, 2, 3)
                .map(integer -> ((Callable<Integer>) (() -> {
                    stream.info("This Service is working on number: " + integer);
                    return integer;
                })
            ))
            .map(asyncTaskExecutor::submit)
            .map(future -> {
                try {
                    return future.get();
                } catch (Exception e) {
                    throw new IllegalStateException(e);
                }
            })
            .collect(Collectors.toList());
            stream.done(collect);

        }, 30000L);
}
```

CHANGELOG.md
### 0.6.0 (2021-02-10)
* improvement: don't throw exception on exceeding 30 chars with a info call but rather shorten it

### 0.2.1 (2021-01-26)
* First feature complete release

### 0.1.0 (2020-12-16)
* Initial release


src/test/java/de/svg/spring_boot_sse_manager/ParallelExecutionTest.java
package de.svg.spring_boot_sse_manager;

import lombok.RequiredArgsConstructor;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.task.AsyncTaskExecutor;

import java.util.List;
import java.util.concurrent.Callable;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@RequiredArgsConstructor
public class ParallelExecutionTest {

    @Autowired
    private AsyncTaskExecutor asyncTaskExecutor;

    //@Test
    public void multiAsyncTest() {
        new SSEStream((SSEStream stream) -> {
            List<Integer> collect = Stream.of(1, 2, 3)
                    .map(
                            integer ->
                                    (
                                            (Callable<Integer>) (
                                                    () -> {
                                                        stream.info("hi " + integer);
                                                        return integer;
                                                    }
                                            )
                                    )
                    )
                    .map(asyncTaskExecutor::submit)
                    .map(
                            future -> {
                                try {
                                    return future.get();
                                } catch (Exception e) {
                                    throw new IllegalStateException(e);
                                }
                            }
                    ).collect(Collectors.toList());
            stream.done(collect);

        }, 30000L);
    }
}


src/main/java/de/svg/spring_boot_sse_manager/dto/Event.java
package de.svg.spring_boot_sse_manager.dto;

import java.util.Date;

import lombok.Getter;
import lombok.Setter;
import org.springframework.http.MediaType;
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter.SseEventBuilder;

/**
 * event for SSEManager.
 */
public class Event {
    @Setter
    private SseEventBuilder builder;

    @Getter
    @Setter
    private Object payload;

    @Getter
    @Setter
    private EventType eventName;

    @Getter
    @Setter
    private Date date;

    public Event(final EventType eventName) {
        buildEvent(eventName, "");
    }

    public Event(final EventType eventName, final Object payload) {
        buildEvent(eventName, payload);
    }

    public SseEmitter.SseEventBuilder get() {
        return builder;
    }

    private void buildEvent(final EventType eventName, final Object payload) {
        setPayload(payload);
        setEventName(eventName);
        setDate(new Date());
        setBuilder(SseEmitter
                .event()
                .name(getEventName().toString())
                .comment(getDate().toString())
                .data(getPayload(), MediaType.APPLICATION_JSON));
    }
}


src/main/java/de/svg/spring_boot_sse_manager/SSEManager.java
package de.svg.spring_boot_sse_manager;

import de.svg.spring_boot_sse_manager.dto.ErrorPayload;
import de.svg.spring_boot_sse_manager.dto.Event;
import de.svg.spring_boot_sse_manager.dto.EventType;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;

import java.io.IOException;
import java.security.InvalidParameterException;
import java.util.Timer;
import java.util.TimerTask;

/**
 * SSEManager.
 */
@Slf4j
public class SSEManager {
    private static final int MAX_INFO_LENGTH = 30;
    private static final int MAX_DEBUG_LENGTH = 255;
    private static final String MESSAGE_TO_LONG_ERROR_MESSAGE = "message is too long, max size is ";

    @Getter
    private final SseEmitter emitter;
    private Integer runningId = 0;
    private Timer timer;

    public SSEManager() {
        emitter = new SseEmitter();
        startHeartBeat();
    }

    public SSEManager(final Long timeout) {
        emitter = new SseEmitter(timeout);
        startHeartBeat();
    }

    public void debug(final String message) {

        if (message.length() < MAX_DEBUG_LENGTH) {
            log.debug("send debug event with message: " + message);
            send(new Event(EventType.DEBUG, message).get());
        } else {
            throw new InvalidParameterException(MESSAGE_TO_LONG_ERROR_MESSAGE + MAX_DEBUG_LENGTH);
        }
    }

    public void info(final String message) {

        if (message.length() < MAX_INFO_LENGTH) {
            log.debug("send info event with message: " + message);
            send(new Event(EventType.INFO, message).get());
        } else {
            throw new InvalidParameterException(MESSAGE_TO_LONG_ERROR_MESSAGE + MAX_INFO_LENGTH);
        }
    }

    public void info() {
        log.debug("send info event");
        send(new Event(EventType.INFO).get());
    }

    public void done(final Object result) {
        log.debug("send done event with object: " + result.toString());
        send(new Event(EventType.DONE, result).get());
        cancelHeartbeat();
        emitter.complete();
    }

    public void error(final Integer status, final Throwable error) {
        final String message = error.getMessage();
        log.error("send error event with message: " + message);
        final ErrorPayload payload = ErrorPayload.builder().status(status).details(message).build();
        send(new Event(EventType.ERROR, payload).get());
        cancelHeartbeat();
        emitter.completeWithError(error);
    }


    public void send(final SseEmitter.SseEventBuilder builder) {
        synchronized (this) {
            try {
                emitter.send(builder.id((++runningId).toString()));
            } catch (IOException ex) {
                log.error(ex.getMessage());
            }
        }
    }

    /**
     * starts heartbeat events every 15 sec to keep sse connection alive.
     */
    private void startHeartBeat() {
        final TimerTask task = new TimerTask() {

            @Override
            public void run() {
                send(new Event(EventType.HEARTBEAT).get());
            }
        };

        timer = new Timer();
        timer.scheduleAtFixedRate(task, 0, 15_000);
    }

    public void cancelHeartbeat() {
        synchronized (this) {
            timer.cancel();
        }
    }

}


src/main/java/de/svg/spring_boot_sse_manager/aop/SSEManagerAspect.java
package de.svg.spring_boot_sse_manager.aop;

import de.svg.spring_boot_sse_manager.SSEManager;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.reflect.MethodSignature;
import org.springframework.stereotype.Component;
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;


import java.lang.reflect.Method;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * SSEManagerAspect.
 */
@Aspect
@Component
public class SSEManagerAspect {

    @Around("@annotation(de.svg.spring_boot_sse_manager.aop.SSEStream)")
    public SseEmitter run(final ProceedingJoinPoint joinPoint) throws Throwable {
        final SSEManager manager;


        final MethodSignature signature = (MethodSignature) joinPoint.getSignature();
        final Method method = signature.getMethod();

        final SSEStream myAnnotation = method.getAnnotation(SSEStream.class);

        if (myAnnotation.timeout().isEmpty()) {
            manager = new SSEManager();
        } else {
            manager = new SSEManager(Long.parseLong(myAnnotation.timeout()));
        }

        final ExecutorService sseMvcExecutor = Executors.newSingleThreadExecutor();

        sseMvcExecutor.execute(() -> {
            final Object proceed;
            try {
                proceed = joinPoint.proceed(new Object[]{manager});
                manager.done(proceed);
            } catch (Throwable throwable) {
                manager.error(500, throwable);
            }
        });

        return manager.getEmitter();
    }
}


src/main/java/de/svg/spring_boot_sse_manager/SSEStream.java
package de.svg.spring_boot_sse_manager;

import de.svg.spring_boot_sse_manager.dto.ErrorPayload;
import de.svg.spring_boot_sse_manager.dto.Event;
import de.svg.spring_boot_sse_manager.dto.EventMessagePayload;
import de.svg.spring_boot_sse_manager.dto.EventType;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.core.task.AsyncTaskExecutor;
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;

import java.io.IOException;
import java.security.InvalidParameterException;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.function.Consumer;

/**
 * SSE stream.
 */
@Slf4j
public class SSEStream extends SseEmitter {
    private static final int MAX_INFO_LENGTH = 30;
    private static final int MAX_DEBUG_LENGTH = 255;
    private static final String MESSAGE_TO_LONG_ERROR_MESSAGE = "message is too long, max size is ";
    private Future<?> callbackFuture;
    private boolean timeoutTriggered;

    private Integer runningId = 0;
    private Timer timer;
    private AsyncTaskExecutor taskExecutor;

    public SSEStream(final Consumer<SSEStream> callback) {
        super();
        onTimeout(this::handleTimeout);
        onCompletion(this::cancelHeartbeat);
        run(callback);
    }

    public SSEStream(final Consumer<SSEStream> callback, final Long timeout) {
        super(timeout);
        onTimeout(this::handleTimeout);
        onCompletion(this::cancelHeartbeat);
        run(callback);
    }


    public SSEStream(final Consumer<SSEStream> callback, final AsyncTaskExecutor taskExecutor) {
        super();
        this.taskExecutor = taskExecutor;
        onTimeout(this::handleTimeout);
        onCompletion(this::cancelHeartbeat);
        run(callback);
    }

    public SSEStream(final Consumer<SSEStream> callback, final Long timeout, final AsyncTaskExecutor taskExecutor) {
        super(timeout);
        this.taskExecutor = taskExecutor;
        onTimeout(this::handleTimeout);
        onCompletion(this::cancelHeartbeat);
        run(callback);
    }

    public void debug(final String message) {
        if (message.length() < MAX_DEBUG_LENGTH) {
            sendEvent(EventType.DEBUG, message);
        } else {
            throw new InvalidParameterException(MESSAGE_TO_LONG_ERROR_MESSAGE + MAX_DEBUG_LENGTH);
        }
    }

    public void debug(final Object message) {
        sendEvent(EventType.DEBUG, message);
    }

    public void info(final String message) {
        if (message.length() > MAX_INFO_LENGTH) {
            log.warn("info call only allows messages up to " + MAX_INFO_LENGTH + " chars and abbreviates the rest.");
        }
        sendEvent(EventType.INFO, StringUtils.abbreviate(message, MAX_INFO_LENGTH));
    }

    public void info(final Object message) {
        sendEvent(EventType.INFO, message);
    }

    public void done(final Object result) {
        log.debug("send done event with object: " + result.toString());
        sendEvent(new Event(EventType.DONE, result).get());
        cancelHeartbeat();
        complete();
    }

    public void error(final Integer status, final Throwable error) {
        final String message = error.getMessage();
        log.error("send error event with message: " + message);
        final ErrorPayload payload = ErrorPayload.builder().status(status).details(message).build();
        sendEvent(new Event(EventType.ERROR, payload).get());
        cancelHeartbeat();
        completeWithError(error);
    }

    private void sendEvent(final EventType type, final String message) {
        log.debug(type.toString() + " event with message: " + message);
        sendEvent(new Event(type, EventMessagePayload.builder().message(message).build()).get());
    }

    private void sendEvent(final EventType type, final Object payload) {
        log.debug(type.toString() + " event with payload: " + payload.toString());
        sendEvent(new Event(type, payload).get());
    }

    public void sendEvent(final SseEmitter.SseEventBuilder builder) {
        synchronized (this) {
            try {
                send(builder.id((++runningId).toString()));
            } catch (IOException ex) {
                log.error("IO Exception" + ex.getMessage());
            }
        }
    }


    private void run(final Consumer<SSEStream> callback) {
        startHeartBeat();
        if (this.taskExecutor != null) {
            callbackFuture = this.taskExecutor.submit(() -> {
                try {
                    callback.accept(this);
                } catch (Throwable throwable) {
                    if (!timeoutTriggered) {
                        error(500, throwable);
                    }
                }
            });
        } else {
            callbackFuture = Executors.newSingleThreadExecutor().submit(() -> {
                try {
                    callback.accept(this);
                } catch (Throwable throwable) {
                    if (!timeoutTriggered) {
                        error(500, throwable);
                    }
                }
            });
        }
    }

    /**
     * starts heartbeat events every 15 sec to keep sse connection alive.
     */
    private void startHeartBeat() {
        final TimerTask task = new TimerTask() {

            @Override
            public void run() {
                sendEvent(new Event(EventType.HEARTBEAT).get());
            }
        };

        timer = new Timer();
        timer.scheduleAtFixedRate(task, 0, 15_000);
    }

    private void cancelHeartbeat() {
        synchronized (this) {
            timer.cancel();
        }
    }

    private void handleTimeout() {
        timeoutTriggered = true;
        log.error("timeout @ " + getTimeout());
        callbackFuture.cancel(true);
    }

}
